internal/alb/tg/targethealth.go

```diff
@@ -0,0 +1,339 @@
+	// Reconcile ensures the target group targets in AWS matches the targets configured in the ingress backend.
+	SyncTargetsForReconciliation(ctx context.Context, t *Targets, desiredTargets []*elbv2.TargetDescription) error
+	RemovePodConditions(ctx context.Context, t *Targets, targets []*elbv2.TargetDescription) error
+	StopReconcilingPodConditionStatus(tgArn string)
+}
+
+// NewTargetHealthController constructs a new target health controller
+func NewTargetHealthController(cloud aws.CloudAPI, store store.Storer, endpointResolver backend.EndpointResolver, client client.Client) TargetHealthController {
+	return &targetHealthController{
+		cloud:            cloud,
+		store:            store,
+		endpointResolver: endpointResolver,
+		client:           client,
+		tgWatches:        make(targetGroupWatches),
+	}
+}
+
+type targetHealthController struct {
+	cloud            aws.CloudAPI
+	store            store.Storer
+	endpointResolver backend.EndpointResolver
+	client           client.Client
+	tgWatches        targetGroupWatches
+	tgWatchesMux     sync.Mutex
+}
+
+// SyncTargetsForReconciliation starts a go routine for reconciling pod condition statuses for the given targets in the background until they are healthy in the target group
+func (c *targetHealthController) SyncTargetsForReconciliation(ctx context.Context, t *Targets, desiredTargets []*elbv2.TargetDescription) error {
+	conditionType := podConditionTypeForIngressBackend(t.Ingress, t.Backend)
+
+	targetsToReconcile, err := c.filterTargetsNeedingReconciliation(conditionType, t, desiredTargets)
+	if err != nil {
+		return err
+	}
+
+	// create, update or remove targetGroupWatch for this target group;
+	// a targetGroupWatch exists as long as there are targets in the target group whose pod condition statuses need to be reconciled;
+	// while the targetGroupWatch exists, a go routine regularly monitors the target health of the targets in the target group and updates the pod condition status for the corresponding pods
+	c.tgWatchesMux.Lock()
+	defer c.tgWatchesMux.Unlock()
+
+	tgWatch, ok := c.tgWatches[t.TgArn]
+	if ok {
+		// targetGroupWatch for this target group already exists
+		if len(targetsToReconcile) == 0 {
+			tgWatch.cancel()
+			delete(c.tgWatches, t.TgArn)
+		} else {
+			tgWatch.mux.Lock()
+			tgWatch.ingress = t.Ingress
+			tgWatch.targetsToReconcile = targetsToReconcile
+			tgWatch.mux.Unlock()
+		}
+	} else {
+		if len(targetsToReconcile) == 0 {
+			return nil
+		}
+
+		// targetGroupWatch for this target group doesn't exist yet -> create it
+		ctx, cancel := context.WithCancel(ctx)
+		tgWatch = &targetGroupWatch{
+			ingress:            t.Ingress,
+			backend:            t.Backend,
+			targetsToReconcile: targetsToReconcile,
+			cancel:             cancel,
+		}
+		c.tgWatches[t.TgArn] = tgWatch
+
+		// start watching target health in target group and updating pod condition status
+		go c.reconcilePodConditionsLoop(ctx, t.TgArn, conditionType, tgWatch)
```

Comment:
this is a bit hacky but works. (it replies on pod been updated and re-trigger the reconcile logic, thus triggers the `if len(targetsToReconcile) == 0`  branch. It should works under current implementation. 
But in my opinion, a better safe approach is like:
```
if len(newTargets) != 0 {
      tgWatch = c.tgWatches[t.TgArn]
      addNewTargetsIntoTgWatch(tgWatch, newTargets)
      if tgWatch.CoroutingCompleted() {
          go tgWatch.NewCoroutingRunIntoCompletionUponTracingNewTargetsHealthy()
      }
}
```

Reply:
I don't think I understand what you mean exactly. You mentioned the `if len(targetsToReconcile) == 0` branch but your code is for the `if len(newTargets) != 0` branch?

Reply:
sorry, i meant that the current code will keep looping and calls describeTargets until we receive a new pod event and triggers an new reconcile(which may not happen in cases like Cx modified the ALB annotation to be malformed).

I have write a pseudo code below: 
TargetHealthController.AddTargets will only be called when adding new targets to targetGroup, which will find the corresponding tgWatch and call tgWatch.AddTargets
```
type TargetHealthController interface {
    AddTargets(ctx context.Context, t* Targets, targets []*elbv2.TargetDescription) error
    RemoveTargets(ctx context.Context, t* Targets, targets []]*elbv2.TargetDescription) error
	StopReconciling(ctx context.Context, tgArn)
}

type TGWatch interface {
    AddTargets(targets []]*elbv2.TargetDescription) error
    RemoveTargets(targets []]*elbv2.TargetDescription) error
    Stop() error
}

type tgWatch struct {
    lbArn string
    targets []*elbv2.TargetDescription
    healthCheckInterval int64
    cloud aws.cloud
    cancel func()

    addTargetsChan chan []*elbv2.TargetDescription
    removeTargetsChan chan []*elbv2.TargetDescription
}

func NewTGWatch(ctx context.Context, lbArn string, targets []*elbv2.TargetDescription, healthCheckInterval int64) *tgWatch {
    ctx, cancel = ctx.WithCancel()
    w := &tgWatch {
        lbArn: lbArn,
        targets: targets,
        healthCheckInterval: healthCheckInterval
        cancel: cancel
    }
    go w.run(ctx)
    return w
}

func (w *tgWatch) AddTargets(targets []*elbv2.TargetDescription) error {
    w.addTargetsChan <- targets
}

func (w *tgWatch) RemoveTargets(targets []]*elbv2.TargetDescription) error {
    w.removeTargetsChan <- targets
}

func (w *tgWatch) run(ctx context.Context) error {
    for {
        var startPull <-chan time.Time
        if len(w.targets) != 0 {
            startPull = time.After(w.healthCheckInterval)
        }
        switch {
            case <-startPull:   # will be ignored if w.targets is empty since switch won't select nil channel
                w.pullOnce()
            case targets := <-addTargetsChan:
                // modify w.targets by adding targets into it, targets already exists don't need to be added.
                addTargetsHandlingDuplications(w.targets, targets)
            case targets :=  <-removeTargetsChan:
                removeTargetsHandlingDuplications(w.targets, targets)
            case <-ctx.Done():
                return
        }
    }
}

func (w *tgWatch) pullOnce(ctx context.Context) {
    targetHealths := w.cloud.DescribeTargetHealth(lbArn, w.targets)
    healthyTargets := filterHealthyTargets(targetHealths)
    w.RemoveTargets(ctx, healthyTargets)

    removePodReadinessGateFromTargets(healthyTargets)
}
```

Reply:
Unfortunately the add/remove part will not really work. If the ingress controller gets killed for any reason after targets have been added to the target group and before the pod condition has been set to `True`, the pod condition would never get `True`. That's why we **must** look at all the desired targets (not only the ones that will be added by a reconciliation loop iteration) and afterwards filter the pods that still need to have their target health checked.

But what we could do instead is to stop the go routine and remove it from the map from within the loop if there are no targets to reconcile anymore.

Reply:
Good point, I haven't considered the controller restart case.
Yeah, we either can stop the go routing or make the loop no-op or block on channels. The point is we should keep track of targets that needs to be checked(and remove targets already turns healthy), are only make the APICall for these targets need to be checked in every loop.

Reply:
@M00nF1sh I changed the communication pattern between the `SyncTargetsForReconciliation` and the `reconcilePodConditionsLoop` to use channels instead of mutexes.